cpp/tests/experimental/louvain_primitives.cu

```diff
@@ -0,0 +1,197 @@
+                     char comma[2] = {0, 0};
+                     printf("keys = ");
+                     for (auto p = b_key; p != e_key; ++p) {
+                       printf("%s%d", comma, *p);
+                       comma[0] = ',';
+                     }
+                     printf("\n");
+                     printf("values = ");
+                     comma[0] = (char)0;
+                     for (auto p = b_value; p != e_value; ++p) {
+                       printf("%s%g", comma, *p);
+                       comma[0] = ',';
+                     }
+                     printf("\n");
+                   });
+
+  if (graph_t::is_multi_gpu) {
+    std::tie(communities, community_in_weight_sums, std::ignore) =
+      cugraph::experimental::sort_and_shuffle_kv_pairs(
+        handle.get_comms(),
+        communities.begin(),
+        communities.end(),
+        community_in_weight_sums.begin(),
+        [] __device__(auto) { return 0; },
+        handle.get_stream());
+  }
+
+  // NOTE: these are populated and shuffled before call to update_by_delta_modularity
+  rmm::device_uvector<weight_t> old_cluster_sum_v(num_verts, handle.get_stream());
+  rmm::device_uvector<weight_t> src_vertex_weight_v(num_verts, handle.get_stream());
+  rmm::device_uvector<weight_t> src_cluster_weight_v(num_verts, handle.get_stream());
+
+  // rmm::device_vector<thrust::tuple<vertex_t, weight_t>> output_v(num_verts);
+  auto output_buffer =
+    cugraph::experimental::allocate_dataframe_buffer<thrust::tuple<vertex_t, weight_t>>(
+      num_verts, handle.get_stream());
+
+  copy_v_transform_reduce_key_aggregated_out_nbr(
+    handle,
+    graph_view,
+    thrust::make_zip_iterator(thrust::make_tuple(
+      old_cluster_sum_v.begin(), src_vertex_weight_v.begin(), src_cluster_weight_v.begin())),
+    cluster_v.begin(),
+    communities.begin(),
+    communities.end(),
+    community_in_weight_sums.begin(),
+    [total_edge_weight, resolution] __device__(
+      auto src, auto neighbor_cluster, auto new_cluster_sum, auto src_info, auto a_new) {
+      auto old_cluster_sum = thrust::get<0>(src_info);
+      auto k_k             = thrust::get<1>(src_info);
+      auto a_old           = thrust::get<2>(src_info);
+
+      weight_t delta_modularity = 2 * (((new_cluster_sum - old_cluster_sum) / total_edge_weight) -
+                                       resolution * (a_new * k_k - a_old * k_k + k_k * k_k) /
+                                         (total_edge_weight * total_edge_weight));
+
+      printf("compute delta_modularity (%d, %d, %g, %g, %g, %g, %g)\n",
+             src,
+             neighbor_cluster,
+             new_cluster_sum,
+             old_cluster_sum,
+             k_k,
+             a_old,
+             a_new);
+
+      return thrust::make_tuple(neighbor_cluster, delta_modularity);
+    },
+    [] __device__(auto p1, auto p2) {
+      printf("reducing (%d, %g) with (%d, %g)\n",
```

Comment:
This print statement gets executed 7613 in the current code displaying junky output.

In this example, each vertex has 2 key/value pairs, so I would expect this to execute 5 times (once for each vertex).  Because of the reduction question above, vertex 0 could execute this multiple times, so perhaps 7 or 8 times.

Reply:
Follow the question I posted in the Slack thrust channel.